src/main/java/org/testng/reporters/FailedReporter.java

```diff
@@ -54,84 +55,50 @@ protected void generateFailureSuite(XmlSuite xmlSuite, ISuite suite, String outp
 
-    synchronized(results) {
       for(Map.Entry<String, ISuiteResult> entry : results.entrySet()) {
         ISuiteResult suiteResult = entry.getValue();
         ITestContext testContext = suiteResult.getTestContext();
 
-        generateXmlTest(suite,
-                        xmlTests.get(testContext.getName()),
+
+        generateXmlTest(testContext.getCurrentXmlTest(),
                         testContext,
                         testContext.getFailedTests().getAllResults(),
                         testContext.getSkippedTests().getAllResults());
       }
-    }
 
     if(null != failedSuite.getTests() && failedSuite.getTests().size() > 0) {
       Utils.writeUtf8File(outputDir, TESTNG_FAILED_XML, failedSuite.toXml());
       Utils.writeUtf8File(suite.getOutputDirectory(), TESTNG_FAILED_XML, failedSuite.toXml());
     }
   }
 
-  /**
-   * Do not rely on this method. The class is used as <code>IReporter</code>.
-   *
-   * @see org.testng.TestListenerAdapter#onFinish(org.testng.ITestContext)
-   * @deprecated this class is used now as IReporter
-   */
-  @Deprecated
-  @Override
-  public void onFinish(ITestContext context) {
-    // Delete the previous file
-//    File f = new File(context.getOutputDirectory(), getFileName(context));
-//    f.delete();
-
-    // Calculate the methods we need to rerun :  failed tests and
-    // their dependents
-//    List<ITestResult> failedTests = getFailedTests();
-//    List<ITestResult> skippedTests = getSkippedTests();
-  }
-
-  private void generateXmlTest(ISuite suite,
-                               XmlTest xmlTest,
+  private void generateXmlTest(XmlTest xmlTest,
                                ITestContext context,
-                               Collection<ITestResult> failedTests,
-                               Collection<ITestResult> skippedTests) {
+                               Set<ITestResult> failedTests,
+                               Set<ITestResult> skippedTests) {
     // Note:  we can have skipped tests and no failed tests
     // if a method depends on nonexistent groups
     if (skippedTests.size() > 0 || failedTests.size() > 0) {
       Set<ITestNGMethod> methodsToReRun = Sets.newHashSet();
 
       // Get the transitive closure of all the failed methods and the methods
       // they depend on
-      Collection[] allTests = new Collection[] {
-          failedTests, skippedTests
-      };
-
-      for (Collection<ITestResult> tests : allTests) {
-        for (ITestResult failedTest : tests) {
-          ITestNGMethod current = failedTest.getMethod();
-          if (current.isTest()) {
-            methodsToReRun.add(current);
-            ITestNGMethod method = failedTest.getMethod();
-            // Don't count configuration methods
-            if (method.isTest()) {
-              List<ITestNGMethod> methodsDependedUpon =
-                  MethodHelper.getMethodsDependedUpon(method, context.getAllTestMethods());
+      Set<ITestResult> allTests = new HashSet<>();
```

Comment:
A solution to avoid a new temp collection could be to add a method and call it once by set.
But your change is good too because it is respecting the initial architecture :)